File: ios/Classes/OnDeviceAiPlugin.swift
import Flutter
import UIKit

public class OnDeviceAiPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private let runner = CoreMLRunner()
  private var eventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(name: "on_device_ai", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "on_device_ai_stream", binaryMessenger: registrar.messenger())
    
    let instance = OnDeviceAiPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    
    switch call.method {
    case "loadModel":
      guard let modelName = args?["modelName"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "modelName is required", details: nil))
        return
      }
      let config = args?["config"] as? [String: Any]
      let useGPU = config?["useGPU"] as? Bool ?? true
      
      do {
        try runner.loadModel(name: modelName, useGPU: useGPU)
        result(nil)
      } catch {
        result(FlutterError(code: "LOAD_FAILED", message: "Failed to load model \(modelName)", details: error.localizedDescription))
      }
      
    case "runText":
      guard let prompt = args?["prompt"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "prompt is required", details: nil))
        return
      }
      
      do {
        let output = try runner.runText(prompt: prompt)
        result(output)
      } catch {
        result(FlutterError(code: "INFERENCE_FAILED", message: "Text inference failed", details: error.localizedDescription))
      }
      
    case "runImage":
      guard let imageBytes = args?["imageBytes"] as? FlutterStandardTypedData else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "imageBytes is required", details: nil))
        return
      }
      
      do {
        let output = try runner.runImage(imageBytes: imageBytes)
        result(output)
      } catch {
        result(FlutterError(code: "INFERENCE_FAILED", message: "Image inference failed", details: error.localizedDescription))
      }
      
    case "startStreamText":
      guard let _ = args?["prompt"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "prompt is required", details: nil))
        return
      }
      // Stream handler logic mock
      DispatchQueue.global().async {
          let words = ["This", "is", "a", "mocked", "stream", "response"]
          for word in words {
              Thread.sleep(forTimeInterval: 0.2)
              DispatchQueue.main.async {
                  self.eventSink?(word)
              }
          }
          DispatchQueue.main.async {
              self.eventSink?(FlutterEndOfEventStream)
          }
      }
      result(nil)
      
    case "dispose":
      runner.dispose()
      result(nil)
      
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
      self.eventSink = events
      return nil
  }
  
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
      self.eventSink = nil
      return nil
  }
}

File: android/src/main/kotlin/com/example/on_device_ai/OnDeviceAiPlugin.kt
package com.example.on_device_ai

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import android.os.Handler
import android.os.Looper

/** OnDeviceAiPlugin */
class OnDeviceAiPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private var runner: TFLiteRunner? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    // Dedicated single-threaded executor so TFLite always runs on the same background thread
    private val ioScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        runner = TFLiteRunner(flutterPluginBinding.applicationContext)

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "on_device_ai")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "on_device_ai_stream")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadModel" -> {
                val modelName = call.argument<String>("modelName")
                    ?: return result.error("INVALID_ARGUMENT", "modelName is required", null)
                val config = call.argument<Map<String, Any>>("config")
                val useGPU = config?.get("useGPU") as? Boolean ?: true

                // Run on background thread - TFLite native init MUST NOT run on main thread
                ioScope.launch {
                    try {
                        runner?.loadModel(modelName, useGPU)
                        mainHandler.post { result.success(null) }
                    } catch (e: Exception) {
                        mainHandler.post { result.error("LOAD_FAILED", "Failed to load model: ${e.message}", null) }
                    }
                }
            }
            "runText" -> {
                val prompt = call.argument<String>("prompt")
                    ?: return result.error("INVALID_ARGUMENT", "prompt is required", null)

                ioScope.launch {
                    try {
                        val output = runner?.runText(prompt)
                        mainHandler.post { result.success(output) }
                    } catch (e: Exception) {
                        mainHandler.post { result.error("INFERENCE_FAILED", "Text inference failed: ${e.message}", null) }
                    }
                }
            }
            "runImage" -> {
                val imageBytes = call.argument<ByteArray>("imageBytes")
                    ?: return result.error("INVALID_ARGUMENT", "imageBytes is required", null)

                ioScope.launch {
                    try {
                        val output = runner?.runImage(imageBytes)
                        mainHandler.post { result.success(output) }
                    } catch (e: Exception) {
                        mainHandler.post { result.error("INFERENCE_FAILED", "Image inference failed: ${e.message}", null) }
                    }
                }
            }
            "startStreamText" -> {
                val prompt = call.argument<String>("prompt")
                    ?: return result.error("INVALID_ARGUMENT", "prompt is required", null)

                ioScope.launch {
                    val words = listOf("This", "is", "a", "mocked", "Android", "stream", "response")
                    words.forEach { word ->
                        delay(200)
                        mainHandler.post { eventSink?.success(word) }
                    }
                    mainHandler.post { eventSink?.endOfStream() }
                }
                result.success(null)
            }
            "dispose" -> {
                ioScope.launch {
                    runner?.dispose()
                    mainHandler.post { result.success(null) }
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        runner?.dispose()
        runner = null
        ioScope.cancel()
    }
}
